File: python3/gas_station.py
"""
Space   : O(n)
Time    : O(n)
"""
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        N = len(gas)
        dp = []

        for i in range(N):
            dp.append(gas[i] - cost[i])

        if sum(dp) < 0:
            return -1

        temp = 0
        ans = -1

        for idx in range(N):
            temp += dp[idx]
            if temp >= 0 and ans == -1:
                ans = idx
            if temp < 0:
                ans = -1
            temp = max(temp, 0)

        return ans

File: golang/gas_station.go
package main

/*
Space   : O(n)
Time    : O(n)
*/

func canCompleteCircuit(gas []int, cost []int) int {
	ans := -1
	temp := []int{}
	n := len(gas)
	s := 0

	for i := 0; i < n; i++ {
		temp = append(temp, gas[i]-cost[i])
		s += gas[i] - cost[i]
	}

	if s < 0 {
		return -1
	}

	cur := 0
	for i, v := range temp {
		cur += v
		if cur >= 0 && ans == -1 {
			ans = i
		}
		if cur < 0 {
			ans = -1
		}
		cur = Max(cur, 0)
	}
	return ans
}
